### ISSUE TITLE
can't run under WSL with zscaler

### ISSUE THREAD

**Issue by edgimar** [rustsec/rustsec#1058]
While installing via `cargo install` works fine, when trying to execute `cargo audit`, the following is printed:
```
$ cargo audit
    Fetching advisory database from `https://github.com/RustSec/advisory-db.git`
error: couldn't fetch advisory database: git operation failed: failed to fetch repo: An IO error occurred when talking to the server
```
The URL is accessible via, for example, wget.  Note that the Zscaler cert is installed and being used in the WSL VM.

**Comment by Shnatsel** [comment]
Could you check if version 0.17.6 works? `cargo install --version=0.17.6`

**Comment by edgimar** [comment]
0.17.6 does appear to work (albeit with about a 7 minute "Updating crates.io index" wait).

**Comment by Shnatsel** [comment]
We switched from using libgit2 to gix+hyper+rustls in 0.18.x. It is likely that this is an incompatibility between Zscaler and either rustls or reqwest. 

The first thing to do to isolate the issue is to take a simple program that downloads a URL with reqwest and try downloading some URLs from behind Zscaler first with native-tls backend, then with the `rustls-tls-native-roots` feature. If it works with native-tls but not rustls, it's a rustls issue; if it doesn't work with either SSL backends, it's a reqwest issue.

**Comment by edgimar** [comment]
Downloading with reqwest works OK, for both `native-tls` and `rustls-tls-native-roots` (in the same WSL shell where cargo-audit fails).

**Comment by Shnatsel** [comment]
Well that's fascinating. 

What about [`gitoxide`](https://github.com/Byron/gitoxide) built with rustls and native certs? The CLI defaults to the curl backend and OpenSSL, you'll need to switch it to `reqwest` + `rustls` (there is a feature for that) *and* also enable the `rustls-tls-native-roots` feature on `reqwest` to match the configuration used by `cargo audit`. I don't think gitoxide has a toggle for `rustls-tls-native-roots` feature, you'll just have to edit the `Cargo.toml` file to enable it.

**Comment by edgimar** [comment]
`gix clone https://github.com/RustSec/advisory-db.git` succeeds when built w/ the following patch:

```patch
diff --git a/Cargo.toml b/Cargo.toml
index 8bb402b77..b7abcee5a 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -34,7 +34,7 @@ default = ["max"]
 ##
 ## As fast as possible, tracing, with TUI progress, progress line rendering with auto-configuration, all transports based on their most mature implementation (HTTP), all `ein` tools, CLI colors and local-time support, JSON output, regex support for rev-specs.
 ## Can be amended with the `http-client-curl-rustls` feature to avoid `openssl` as backend.
-max = ["max-control", "fast", "gitoxide-core-blocking-client", "http-client-curl"]
+max = ["max-control", "fast", "gitoxide-core-blocking-client", "http-client-curl-rustls"]
 
 ## Like `max`, but only Rust is allowed.
 ##
@@ -50,7 +50,7 @@ max-control = ["tracing", "fast-safe", "pretty-cli", "gitoxide-core-tools-query"
 ## All of the good stuff, with less fanciness for smaller binaries.
 ##
 ## As fast as possible, progress line rendering, all transports based on their most mature implementation (HTTP), all `ein` tools, CLI colors and local-time support, JSON output.
-lean = ["fast", "tracing", "pretty-cli", "http-client-curl", "gitoxide-core-tools-query", "gitoxide-core-tools-corpus", "gitoxide-core-tools", "gitoxide-core-blocking-client", "prodash-render-line" ]
+lean = ["fast", "tracing", "pretty-cli", "http-client-curl-rustls", "gitoxide-core-tools-query", "gitoxide-core-tools-corpus", "gitoxide-core-tools", "gitoxide-core-blocking-client", "prodash-render-line" ]
 
 ## The smallest possible build, best suitable for small single-core machines.
 ##
diff --git a/gix-transport/Cargo.toml b/gix-transport/Cargo.toml
index 155d6acc9..a0285e47b 100644
--- a/gix-transport/Cargo.toml
+++ b/gix-transport/Cargo.toml
@@ -78,7 +78,8 @@ base64 = { version = "0.21.0", optional = true }
 curl = { version = "0.4", optional = true }
 
 # for http-client-reqwest
-reqwest = { version = "0.11.12", optional = true, default-features = false, features = ["blocking"] }
+reqwest = { version = "0.11.12", optional = true, default-features = false, features = ["blocking", "rustls-tls-native-roots"] }
+
 
 ## If used in conjunction with `async-client`, the `connect()` method will become available along with supporting the git protocol over TCP,
 ## where the TCP stream is created using this crate.
```

Also, just to confirm that the failure with cargo-audit wasn't a fluke, I reinstalled 0.18.3, and am still seeing the same "IO error".

**Comment by Shnatsel** [comment]
Strange. I don't think we're doing anything particularly interesting with our `reqwest` configuration. It might be that our `gix` is slightly outdated, we're not tracking the latest upstream right now. Let me see if upgrading will do anything.

#1053 might also help debug this.

**Comment by Shnatsel** [comment]
I suggest also opening a support ticket with zscaler developers. This is a proprietary and paywalled tool that we do not have access to, so it is going to be quite tricky for us to debug any issues related to it.

**Comment by Shnatsel** [comment]
Could you try #1063 and see if that produces a more informative error message?

**Comment by emanuel-minetti** [comment]
I'm having quite a similar issue (and maybe a reason). Developing on Windows and having freshly installed (and used) I got the error messages like
```
Scanning Cargo.lock for vulnerabilities (192 crate dependencies)
error: couldn't check if the package is yanked: registry: Failed to retrieve actix-codec from crates.io index: registry: request could not be completed in the allotted timeframe
```
followed by a lot of output (one for each dependency).

I suspected a connection to Windows, so I tried WSL. Same result.

After that I tried installing 1.17.6
`cargo install cargo-audit --force --version=0.17.6`
and it took nearly 10 minutes but it worked.

It gave me one warning for my project which disappeared after running `cargo update`.

I think I found one (or the) reason for this:
When I use 1.18.3 I'll find a file
```
advisory-db..lock
```
but when I use 1.17.6 I find a file
```
advisory-db.lock
```
beyond `.cargo`.

Notice the extra dot! It happened on normal Windows (11) and WSL. Hope that helps finding the (real) reason and that it fixes my problem as well .`

**Comment by Shnatsel** [comment]
Yeah, the `advisory-db..lock` is my goof added in 0.18.x. I don't think it's the culprit - I have it on my Linux system too and `cargo audit` works fine.

**Comment by emanuel-minetti** [comment]
Ok, but if I use 1.18.3 there is no `advisory-db.lock`. And it seems that `cargo-audit/src/auditor.rs` awaits the file to be there.

**Comment by asomers** [comment]
I am also seeing this problem on Cirrus CI's platform with FreeBSD.  Every CI run that uses cargo-audit 0.18.3 (which is in the package repository for FreeBSD 14) fails with this same error.  But CI runs that use 0.17.6 (which is in the package repository for FreeBSD 13.2) succeeds.  I cannot reproduce the problem on the bench.  Probably, Cirrus CI is using something like cloudflare or zscaler.  Some example failures:
* https://cirrus-ci.com/task/6480078427652096
* https://cirrus-ci.com/task/5822941653368832
* https://cirrus-ci.com/task/6730209940996096

**Comment by Shnatsel** [comment]
I have just published version 0.19.0. If it doesn't fix this issue, then at least it should provide a more informative error message.

It needs to be paired with the latest `rustsec` crate for you to get these improvements, so please either install from `cargo` or make sure your distribution package uses `rustsec` v0.28.5.

**Comment by emanuel-minetti** [comment]
Fixed my problem.

(Whereas the weird file `advisory-db..lock` still shows up.)

THANKS A LOT

**Comment by Shnatsel** [comment]
Thanks! I'll wait for one more report of this being fixed before closing, just to make sure.

It may have been fixed by https://github.com/EmbarkStudios/tame-index/pull/47 so please check that you have `tame-index` 0.9.3 or later in your dependency tree. Running `cargo install cargo-audit --force` today should get a recent enough version.

This fix is actually not included in the 0.19.0 binary of `cargo audit` on Github releases.

**Comment by emanuel-minetti** [comment]
Before my comment I ran `cargo install cargo-audit --force`  (running `cargo audit --version` gives `0.19.0`) and it solved **my** problem.
I verified that `tame-index` 0.9.3 is installed.

Thanks again.

**Comment by asomers** [comment]
And FTR cargo-audit 0.19.0 works in Cirrus CI, too.

**Comment by jgpaiva** [comment]
Unfortunately I think the originally reported problem (IO error when running in enterprise environment under a proxy that intercepts TLS) is still happening. Under cloudflare gateway, I see:
```
╰─❯ cargo audit
    Fetching advisory database from `https://github.com/RustSec/advisory-db.git`
error: couldn't fetch advisory database: git operation failed: failed to prepare fetch: An IO error occurred when talking to the server

╰─❯ cargo audit --version
cargo-audit-audit 0.19.0

```

**Comment by Shnatsel** [comment]
@jgpaiva I have just shipped v0.20.0 that makes the dependencies with the necessary fixes a hard requirement. Please try it, and if it doesn't work, please open a new issue with the steps to reproduce it.
